### ISSUE TITLE
[bug] FileNotFoundError when insufficient permissions to create .conan directory.

### ISSUE THREAD

**Issue by elektito** [conan-io/conan#7281]
### Environment Details (include every applicable attribute)
  * Operating system: Ubuntu 18.04
  * Conan version: 1.24.0 (also tested with current develop branch)
  * Python version: 3.6.9

### Steps to reproduce
Run this as a non-root user:

```CONAN_USER_HOME=/ conan remote add foobar http://example.com```

You get an error ending with "No such file or directory: '/.conan/conan.conf'" (see logs), which is misleading. There should be a permission error.

### Logs
```
Traceback (most recent call last):
  File "/home/mostafaw/source/conan/venv/bin/conan", line 11, in <module>
    load_entry_point('conan==1.27.0.dev0', 'console_scripts', 'conan')()
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/conan.py", line 7, in run
    main(sys.argv[1:])
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/command.py", line 2184, in main
    conan_api, _, _ = Conan.factory()
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/conan_api.py", line 224, in factory
    return cls(), None, None
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/conan_api.py", line 236, in __init__
    migrator = ClientMigrator(self.cache_folder, Version(client_version), self.out)
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/migrations.py", line 27, in __init__
    self.cache = ClientCache(cache_folder, out)
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/cache/cache.py", line 78, in __init__
    self._store_folder = self.config.storage_path or self.cache_folder
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/cache/cache.py", line 150, in config
    self.initialize_config()
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/client/cache/cache.py", line 240, in initialize_config
    save(self.conan_conf_path, normalize(get_default_client_conf()))
  File "/home/mostafaw/source/conan/venv/lib/python3.6/site-packages/conan-1.27.0.dev0-py3.6.egg/conans/util/files.py", line 194, in save
    with open(path, "wb") as handle:
FileNotFoundError: [Errno 2] No such file or directory: '/.conan/conan.conf'
```
### More explanation

The problem is in the `conans.util.files.save` function which blindly silences all errors raised by `os.makedirs`. I believe the intention has been to silence the "file exists" errors. This can better be achieved by passing `exist_ok=True` to `os.makedirs` (supported from Python 3.2 upwards). 

Another way to fix the issue is using the already existing `mkdir` function in the same module, which tests for the directory's existence first. I personally don't think that's the best idea, since the way it's implemented (check first, then create) is a race condition. I think that function can be re-written to either use `exist_ok` (if minimum Python 3.2 support is acceptable), or use a try block. If first option is acceptable, the function becomes a bit pointless and I would personally go with removing it entirely and replacing all invocations with plain `os.makedir(path, exist_ok=True)`.

I have already created a small patch (by removing the try block and adding the exist_ok option), but the contributors guide says I'd better open an issue first to gather feedback.

**Comment by memsharded** [comment]
Hi @elektito 

Thanks for the report.  Let's say the rule is that a stack trace in the output is always a bug, and it should better be a controlled error. Sure, please send a PR referencing this issue, and we will see what happens in CI. Take into account that we are still running Python2.7 in CI, even if deprecated and soon to be removed, but at the moment it should still work.

**Comment by elektito** [comment]
So I created a PR which I believe should fix the original problem I mentioned (wrong error). But it still gives a stack trace. I'm not familiar enough with the Conan code base to know how to fix that. If you have any pointers about that, I'm willing to see if I can update the PR.

**Comment by elektito** [comment]
Also, looks like the tests failed. At a glance, the failure seems to be something that had been caught and handled by the removed try block. Not sure if the test is wrong, or the my patch. I'll investigate.

**Comment by elektito** [comment]
So apparently empty directory names were being passed and expected to be ignored. I added a check to not call `mkdir` for those. I wanted to add the check to `mkdir` at first (it would have been a bit prettier perhaps), but I thought that would be pushing the extent of the issue and affecting other pieces of code.

Also, however many times I sign that CLA, it still tells me I need to sign it.

**Comment by memsharded** [comment]
>So apparently empty directory names were being passed and expected to be ignored. I added a check to not call mkdir for those. I wanted to add the check to mkdir at first (it would have been a bit prettier perhaps), but I thought that would be pushing the extent of the issue and affecting other pieces of code.

Yes, this is the existing behavior and seems cannot be changed, the ``save()`` and ``save_append()`` can receive a bare filename, and will save the file in the current dir. Changing this behavior would be breaking for many users.

> Also, however many times I sign that CLA, it still tells me I need to sign it.

This is typically due to some mismatch between the user/email you are using to do the commits and the one you are using in your Github login.

**Comment by elektito** [comment]
> This is typically due to some mismatch between the user/email you are using to do the commits and the one you are using in your Github login.

This should be the problem. I'll fix it.

**Comment by elektito** [comment]
Just updated the PR and my CLA is now recognized. I think the tests were passing before this, but we'll see if my new email has any effect on the results! :grin: 

What about the other issue, that this still gives a stack trace (albeit the correct one this time)? Should that be fixed in another issue/PR?

**Comment by memsharded** [comment]
Closed, check https://github.com/conan-io/conan/pull/7284#issuecomment-1367259811, considered solved in 2.0. Thanks!
